README.md

# R/`drtmle`

[![Travis-CI Build
Status](https://travis-ci.org/benkeser/drtmle.svg?branch=master)](https://travis-ci.org/benkeser/drtmle)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/benkeser/drtmle?branch=master&svg=true)](https://ci.appveyor.com/project/benkeser/drtmle)
[![Coverage
Status](https://img.shields.io/codecov/c/github/benkeser/drtmle/master.svg)](https://codecov.io/github/benkeser/drtmle?branch=master)
[![CRAN](http://www.r-pkg.org/badges/version/drtmle)](http://www.r-pkg.org/pkg/drtmle)
[![CRAN
downloads](https://cranlogs.r-pkg.org/badges/drtmle)](https://CRAN.R-project.org/package=drtmle)
[![Project Status: Active - The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![MIT
license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)
[![DOI](https://zenodo.org/badge/75324341.svg)](https://zenodo.org/badge/latestdoi/75324341)

> Nonparametric estimators of the average treatment effect with
> doubly-robust confidence intervals and hypothesis tests

**Author:** [David
Benkeser](https://www.sph.emory.edu/faculty/profile/#!dbenkes)

------------------------------------------------------------------------

## Description

`drtmle` is an R package that computes marginal means of an outcome
under fixed levels of a treatment. The package computes targeted minimum
loss-based (TMLE) estimators that are doubly robust, not only with
respect to consistency, but also with respect to asymptotic normality,
as discussed in [Benkeser, et al.
(2017)](https://academic.oup.com/biomet/article-abstract/104/4/863/4554445).
This property facilitates construction of doubly-robust confidence
intervals and hypothesis tests.

The package additionally includes methods for computing valid confidence
intervals for an inverse probability of treatment weighted (IPTW)
estimator of the average treatment effect when the propensity score is
estimated via super learning, as discussed in [van der Laan
(2014)](https://www.degruyter.com/downloadpdf/j/ijb.2014.10.issue-1/ijb-2012-0038/ijb-2012-0038.pdf).

------------------------------------------------------------------------

## Installation

Install the current stable release from
[CRAN](https://cran.r-project.org/) via

``` r
install.packages("drtmle")
```

A developmental release may be installed from GitHub via
[`devtools`](https://www.rstudio.com/products/rpackages/devtools/) with:

``` r
devtools::install_github("benkeser/drtmle")
```

------------------------------------------------------------------------

## Usage

### Doubly-robust inference for the average treatment effect

Suppose the data consist of a vector of baseline covariates (`W`), a
multi-level treatment assignment (`A`), and a continuous or
binary-valued outcome (`Y`). The function `drtmle` may be used to
estimate $E[E(Y \mid A = a_0, W)]$ for user-selected values of $a_0$
(via option `a_0`). The resulting targeted minimum loss-based estimates
are doubly robust with respect to both consistency and asymptotic
normality. The function computes doubly robust covariance estimates that
can be used to construct doubly robust confidence intervals for marginal
means and contrasts between means. A simple example on simulated data is
shown below. We refer users to [the
vignette](https://benkeser.github.io/drtmle/articles/using_drtmle.html)
for more information and further examples.

``` r
# load packages
library(drtmle)
#> drtmle: TMLE with doubly robust inference
#> Version: 1.1.1
library(SuperLearner)
#> Loading required package: nnls
#> Loading required package: gam
#> Loading required package: splines
#> Loading required package: foreach
#> Loaded gam 1.20.1
#> Super Learner
#> Version: 2.0-28
#> Package created on 2021-05-04

# simulate simple data structure
set.seed(12345)
n <- 200
W <- data.frame(W1 = runif(n,-2,2), W2 = rbinom(n,1,0.5))
A <- rbinom(n, 1, plogis(-2 + W$W1 - 2*W$W1*W$W2))
Y <- rbinom(n, 1, plogis(-2 + W$W1 - 2*W$W1*W$W2 + A))

# estimate the covariate-adjusted marginal mean for A = 1 and A = 0
# here, we do not properly estimate the propensity score
fit1 <- drtmle(W = W, A = A, Y = Y, # input data
               a_0 = c(0, 1), # return estimates for A = 0 and A = 1
               SL_Q = "SL.npreg", # use kernel regression for E(Y | A = a, W)
               glm_g = "W1 + W2", # use misspecified main terms glm for E(A | W)
               SL_Qr = "SL.npreg", # use kernel regression to guard against
                                   # misspecification of outcome regression
               SL_gr = "SL.npreg", # use kernel regression to guard against
                                  # misspecification of propensity score
               returnModels = TRUE # for visualizing fits later
              )
# print the output
fit1
#> $est
#>            
#> 0 0.1752271
#> 1 0.2866095
#> 
#> $cov
#>              0            1
#> 0 9.039683e-04 4.591974e-05
#> 1 4.591974e-05 8.823850e-03

# get confidence intervals for marginal means
# truth is E[Y(1)] = 0.29, E[Y(0)] = 0.15
ci_fit1 <- ci(fit1)
# print the output
ci_fit1
#> $drtmle
#>     est   cil   ciu
#> 0 0.175 0.116 0.234
#> 1 0.287 0.102 0.471

# get confidence intervals for ate
# truth is E[Y(1)] - E[Y(0)] = 0.14
ci_ate1 <- ci(fit1, contrast = c(-1, 1))
# print the output
ci_ate1
#> $drtmle
#>                   est    cil   ciu
#> E[Y(1)]-E[Y(0)] 0.111 -0.081 0.304
```

This method requires estimation of additional univariate regressions to
ensure doubly robust confidence intervals and hypothesis tests. The
method for estimation are input via `SL.Qr` and `SL.gr` or `glm.Qr` and
`glm.gr` if parametric models are desired). These additional fits can be
visualized by the `plot` method for `drtmle`.

``` r
layout(t(1:3))
plot(fit1, ask = FALSE)
```

### Inference for super learner-based IPTW

The package additionally includes a function for computing valid
confidence intervals about an inverse probability of treatment weight
(IPTW) estimator when super learning is used to estimate the propensity
score.

``` r
# fit iptw
fit2 <- adaptive_iptw(Y = Y, A = A, W = W, a_0 = c(0, 1),
                      SL_g = c("SL.glm", "SL.mean", "SL.step.interaction"),
                      SL_Qr = "SL.npreg")
#> Loading required package: nloptr
# print the output
fit2
#> $est
#>            
#> 0 0.1734251
#> 1 0.2438025
#> 
#> $cov
#>              0            1
#> 0 8.607877e-04 8.982892e-05
#> 1 8.982892e-05 2.308177e-02

# compute a confidence interval for margin means
ci_fit2 <- ci(fit2)
# print the output
ci_fit2
#> $iptw_tmle
#>     est    cil   ciu
#> 0 0.173  0.116 0.231
#> 1 0.244 -0.054 0.542

# compute a confidence interval for the ate
ci_ate2 <- ci(fit2, contrast = c(-1, 1))
# print the output
ci_ate2
#> $iptw_tmle
#>                  est    cil   ciu
#> E[Y(1)]-E[Y(0)] 0.07 -0.232 0.373
```

------------------------------------------------------------------------

## Issues

If you encounter any bugs or have any specific feature requests, please
[file an issue](https://github.com/benkeser/drtmle/issues).

------------------------------------------------------------------------

## Citation

After using the `drtmle` R package, please cite the following:

    @Manual{drtmlepackage,
      title = {drtmle: Doubly-Robust Nonparametric Estimation and Inference},
      author = {David Benkeser},
      note = {R package version 1.0.0},
      doi = {10.5281/zenodo.844836}
    }

    @article{benkeser2017improved,
      year  = {2017},
      author = {Benkeser, David C and Carone, Marco and van der Laan, Mark J
        and Gilbert, Peter B},
      title = {Doubly-robust nonparametric inference on the average
        treatment effect},
      journal = {Biometrika},
      volume = {104}, number = {4},
      pages = {863–880},
      doi = {10.1093/biomet/asx053}
    }

------------------------------------------------------------------------

## License

© 2016- [David C.
Benkeser](https://www.sph.emory.edu/faculty/profile/#!dbenkes)

The contents of this repository are distributed under the MIT license.
See below for details:

    The MIT License (MIT)

    Copyright (c) 2016- David C. Benkeser

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.


NEWS.md
# drtmle 1.1.2

* Version 1.1.2 released on GitHub and CRAN.
* Major changes
	* fixed bug in standard error calculation when user inputs estimates

# drtmle 1.1.1

* Version 1.1.1 released on GitHub and CRAN.
* Major changes
	* fixed bug in how multi-level treatment estimates are computed
* Minor changes
	* fixed small bugs that caused errors predicting from some super learner
	wrapper functions.
	* removed `earth` dependency to fix build errors

# drtmle 1.1.0

* Version 1.1.0 released on GitHub and CRAN.
* Major changes
	* add support for averaging over multiple super learner runs
	* add support for partially/fully cross-validated standard error estimates
	* default for `stratify` argument changed to `FALSE`
	* major additions to vignette documenting new features
* Minor changes
	* `future` parallel backend must be specified externally by user
	* accordingly, some dependencies have been updated

# drtmle 1.0.5

As of January 2019:
* Version 1.0.5 released on GitHub and CRAN.
* Addition of `snow` and `data.table` packages to suggests to quell re-occurring
   warnings and errors on CRAN builds.
* Added `nSL` option for fitting and averaging multiple super learners as part
   of the estimation procedure. 
* Added `adapt_g` option for outcome-adaptive propensity score fitting. 

# drtmle 1.0.4

As of August 2019:
* Added minor touch-ups and link fixes to documentation and vignettes.
* Improved how slots are ordered upon being included in the return object.

As of July 2019:
* Version 1.0.4.9001 released on GitHub.
* Removed dependency on `plyr` package.
* Added option to average over repeated Super Learner fits.

As of December 18, 2018:
* Version 1.0.4 released on GitHub and CRAN.
* Resolved issues arising from `returnModels` option when users input nuisance
   parameters.
* Added option to bypass `future` parallelization calls for easier debugging.
* Fixed bugs in standard TMLE implementation -- namely, more robust fluctuations
   and corrected variance estimators.

# drtmle 1.0.3

As of July 2, 2018:
* Version 1.0.3 released on GitHub and CRAN.
* Fixed warnings on CRAN builds.

# drtmle 1.0.2

As of February 5, 2018:
* Version 1.0.2 released on GitHub and CRAN.
* Replaced `foreach` parallelization with `future`.
* Included more robust Super Learner methods.
* Fixed test to pass build with long doubles removed.
* Accommodated returning estimated influence functions with `drtmle()` fit for
   power users.
* Incorporated minor documentation corrections and updates.

As of December 11, 2017:
* Version 1.0.2.9000 released on GitHub.
* More robust convex combination SuperLearner implemented.

# drtmle 1.0.0

As of August 17, 2017:
* Version 1.0.0 released on CRAN.
* Version 1.0.0.9000 released on GitHub.

As of August 15, 2017:
* Version 1.0.0 ready for CRAN release.

# drtmle 0.0.1

As of April 05, 2017:
* The first public release of this package (v. 0.0.1) is made available on
   GitHub.


R/confint.R
#' Compute confidence intervals for drtmle and adaptive_iptw@
#' @param ... Arguments to be passed to method
#' @export
ci <- function(...) {
  UseMethod("ci")
}

#' Confidence intervals for drtmle objects
#'
#' @param object An object of class \code{"drtmle"}
#' @param est A vector indicating for which estimators to return a
#'  confidence interval. Possible estimators include the TMLE with doubly robust
#'  inference (\code{"drtmle"}, recommended), the AIPTW with additional
#'  correction for misspecification (\code{"aiptw_c"}, not recommended), the
#'  standard TMLE (\code{"tmle"}, recommended only for comparison to "drtmle"),
#'  the standard AIPTW (\code{"aiptw"}, recommended only for comparison to
#'  "drtmle"), and G-computation (\code{"gcomp"}, not recommended).
#' @param level The nominal coverage probability of the desired confidence
#'  interval (should be between 0 and 1). Default computes 95\% confidence
#'  intervals.
#' @param contrast Specifies the parameter for which to return confidence
#'  intervals. If \code{contrast=NULL}, then confidence intervals for the
#'  marginal means are computed. If instead, \code{contrast} is a numeric vector
#'  of ones, negative ones, and zeros to define linear combinations of the
#'  various means (e.g., to estimate an average treatment effect, see example).
#'  Finally, \code{contrast} can be a list with named functions \code{f},
#'  \code{f_inv}, \code{h}, and \code{fh_grad}. The first two functions should
#'  take as input argument \code{eff}. Respectively, these specify which
#'  transformation of the effect measure to compute the confidence interval for
#'  and the inverse transformation to put the confidence interval back on the
#'  original scale. The function \code{h} defines the contrast to be estimated
#'  and should take as input \code{est}, a vector of the same length as
#'  \code{object$a_0}, and output the desired contrast. The function
#'  \code{fh_grad} is the gradient of the function \code{h}. See examples and
#'  vignette for more information.
#' @param ... Other options (not currently used).
#' @importFrom stats qnorm
#' @export
#' @method ci drtmle
#' @return An object of class \code{"ci.drtmle"} with point estimates and
#'  confidence intervals of the specified level.
#'
#' @examples
#' # load super learner
#' library(SuperLearner)
#' # simulate data
#' set.seed(123456)
#' n <- 100
#' W <- data.frame(W1 = runif(n), W2 = rnorm(n))
#' A <- rbinom(n, 1, plogis(W$W1 - W$W2))
#' Y <- rbinom(n, 1, plogis(W$W1 * W$W2 * A))
#' \donttest{
#' # fit drtmle with maxIter = 1 to run fast
#' fit1 <- drtmle(
#'   W = W, A = A, Y = Y, a_0 = c(1, 0),
#'   family = binomial(),
#'   stratify = FALSE,
#'   SL_Q = c("SL.glm", "SL.mean"),
#'   SL_g = c("SL.glm", "SL.mean"),
#'   SL_Qr = "SL.npreg",
#'   SL_gr = "SL.npreg", maxIter = 1
#' )
#'
#' # get confidence intervals for each mean
#' ci_mean <- ci(fit1)
#'
#' # get confidence intervals for ATE
#' ci_ATE <- ci(fit1, contrast = c(1, -1))
#'
#' # get confidence intervals for risk ratio by
#' # computing CI on log scale and back-transforming
#' myContrast <- list(
#'   f = function(eff) {
#'     log(eff)
#'   },
#'   f_inv = function(eff) {
#'     exp(eff)
#'   },
#'   h = function(est) {
#'     est[1] / est[2]
#'   },
#'   fh_grad = function(est) {
#'     c(1 / est[1], -1 / est[2])
#'   }
#' )
#' ci_RR <- ci(fit1, contrast = myContrast)
#' }
ci.drtmle <- function(object, est = c("drtmle"), level = 0.95,
                      contrast = NULL, ...) {
  if (!inherits(object, "drtmle")) {
    stop("ci only works with drtmle objects")
  }
  out <- vector(mode = "list", length = length(est))
  names(out) <- est
  # if no contrast then return an CI for each
  # covariate-adjusted mean
  if (is.null(contrast)) {
    for (i in seq_along(est)) {
      out[[i]] <- matrix(NA, nrow = length(object$a_0), ncol = 3)
      for (j in seq_along(object$a_0)) {
        out[[i]][j, ] <-
          rep(object[[est[i]]]$est[j], 3) +
          stats::qnorm(c(0.5, (1 - level) / 2, (1 + level) / 2)) *
            rep(sqrt(object[[est[i]]]$cov[j, j]), 3)
      }
      row.names(out[[i]]) <- object$a_0
      colnames(out[[i]]) <- c("est", "cil", "ciu")
    }
  } else if (is.numeric(contrast)) {
    # check that contrast came in correctly
    if (length(contrast) != length(object$a_0)) {
      stop("length of contrast vector not equal to length of a_0")
    }
    if (!all(contrast %in% c(-1, 1, 0))) {
      stop("contrast should only be -1, 1, or 0")
    }
    for (i in seq_along(est)) {
      out[[i]] <- matrix(NA, nrow = 1, ncol = 3)
      g <- matrix(contrast, nrow = length(contrast))
      p <- matrix(object[[est[i]]]$est, nrow = length(contrast))
      v <- object[[est[i]]]$cov
      thisC <- t(g) %*% p
      thisSe <- sqrt(t(g) %*% v %*% g)
      out[[i]][1, ] <- rep(thisC, 3) +
        stats::qnorm(c(0.5, (1 - level) / 2, (1 + level) / 2)) *
          rep(thisSe, 3)
      # E[Y(a_0[1])] - E[Y(a_0[2])]
      indMinus <- which(contrast == -1)
      indPlus <- which(contrast == 1)
      plusTerms <- minusTerms <- ""
      if (length(indMinus) > 0) {
        minusTerms <- paste0("E[Y(", object$a_0[indMinus], ")]")
      }
      if (length(indPlus) > 0) {
        plusTerms <- paste0("E[Y(", object$a_0[indPlus], ")]")
      }
      thisName <- paste0(
        paste0(plusTerms, collapse = "+"),
        ifelse(length(indMinus) > 0, "-", ""),
        paste0(minusTerms, collapse = "-")
      )

      row.names(out[[i]]) <- thisName
      colnames(out[[i]]) <- c("est", "cil", "ciu")
    }
  } else if (is.list(contrast)) {
    if (!all(c("f", "f_inv", "h", "fh_grad") %in% names(contrast))) {
      stop("some function missing in contrast. see ?ci.drtmle for help.")
    }
    for (i in seq_along(est)) {
      out[[i]] <- matrix(NA, nrow = 1, ncol = 3)
      thisC <- do.call(contrast$h, args = list(est = object[[est[i]]]$est))
      f_thisC <- do.call(contrast$f, args = list(eff = thisC))

      grad <- matrix(do.call(
        contrast$fh_grad,
        args = list(est = object[[est[i]]]$est)
      ), nrow = length(object$a_0))
      v <- object[[est[i]]]$cov
      thisSe <- sqrt(t(grad) %*% v %*% grad)
      transformCI <- rep(f_thisC, 3) +
        stats::qnorm(c(0.5, (1 - level) / 2, (1 + level) / 2)) *
          rep(thisSe, 3)
      out[[i]][1, ] <- do.call(contrast$f_inv, args = list(eff = transformCI))
      row.names(out[[i]]) <- c("user contrast")
      colnames(out[[i]]) <- c("est", "cil", "ciu")
    }
  }
  class(out) <- "ci.drtmle"
  return(out)
}

#' Confidence intervals for adaptive_iptw objects
#'
#' Estimate confidence intervals for objects of class \code{"adaptive_iptw"}
#'
#' @param object An object of class \code{"adaptive_iptw"}
#' @param est A vector indicating for which estimators to return a
#' confidence interval. Possible estimators include the TMLE IPTW
#' (\code{"iptw_tmle"}, recommended), the one-step IPTW
#' (\code{"iptw_os"}, not recommended), the standard IPTW
#' (\code{"iptw"}, recommended only for comparison to the other two estimators).
#' @param level The nominal coverage probability of the desired confidence
#'  interval (should be between 0 and 1). Default computes 95\% confidence
#'  intervals.
#' @param contrast Specifies the parameter for which to return confidence
#'  intervals. If \code{contrast=NULL}, then confidence intervals for the
#'  marginal means are computed. If instead, \code{contrast} is a numeric vector
#'  of ones, negative ones, and zeros to define linear combinations of the
#'  various means (e.g., to estimate an average treatment effect, see example).
#'  Finally, \code{contrast} can be a list with named functions \code{f},
#'  \code{f_inv}, \code{h}, and \code{fh_grad}. The first two functions should
#'  take as input argument \code{eff}. Respectively, these specify which
#'  transformation of the effect measure to compute the confidence interval for
#'  and the inverse transformation to put the confidence interval back on the
#'  original scale. The function \code{h} defines the contrast to be estimated
#'  and should take as input \code{est}, a vector of the same length as
#'  \code{object$a_0}, and output the desired contrast. The function
#'  \code{fh_grad} is the gradient of the function \code{h}. See examples and
#'  vignette for more information.
#' @param ... Other options (not currently used).
#' @export
#' @method ci adaptive_iptw
#' @return An object of class \code{"ci.adaptive_iptw"} with point estimates and
#'  confidence intervals of the specified level.
#'
#' @examples
#' # load super learner
#' library(SuperLearner)
#' # fit adaptive_iptw
#' set.seed(123456)
#' n <- 200
#' W <- data.frame(W1 = runif(n), W2 = rnorm(n))
#' A <- rbinom(n, 1, plogis(W$W1 - W$W2))
#' Y <- rbinom(n, 1, plogis(W$W1 * W$W2 * A))
#'
#' fit1 <- adaptive_iptw(
#'   W = W, A = A, Y = Y, a_0 = c(1, 0),
#'   SL_g = c("SL.glm", "SL.mean", "SL.step"),
#'   SL_Qr = "SL.glm"
#' )
#'
#' # get confidence intervals for each mean
#' ci_mean <- ci(fit1)
#'
#' # get confidence intervals for ATE
#' ci_ATE <- ci(fit1, contrast = c(1, -1))
#'
#' # get confidence intervals for risk ratio
#' # by inputting own contrast function
#' # this computes CI on log scale and back transforms
#' myContrast <- list(
#'   f = function(eff) {
#'     log(eff)
#'   },
#'   f_inv = function(eff) {
#'     exp(eff)
#'   },
#'   h = function(est) {
#'     est[1] / est[2]
#'   },
#'   fh_grad = function(est) {
#'     c(1 / est[1], -1 / est[2])
#'   }
#' )
#' ci_RR <- ci(fit1, contrast = myContrast)
ci.adaptive_iptw <- function(object, est = c("iptw_tmle"), level = 0.95,
                             contrast = NULL, ...) {
  if (any(est == "iptw")) {
    stop("Theory does not support inference for naive IPTW with super learner.")
  }
  if (!inherits(object, "adaptive_iptw")) {
    stop("ci only works with adaptive_iptw objects")
  }
  out <- vector(mode = "list", length = length(est))
  names(out) <- est
  # if no contrast then return an CI for each
  # covariate-adjusted mean
  if (is.null(contrast)) {
    for (i in seq_along(est)) {
      out[[i]] <- matrix(NA, nrow = length(object$a_0), ncol = 3)
      for (j in seq_along(object$a_0)) {
        out[[i]][j, ] <-
          rep(object[[est[i]]]$est[j], 3) +
          stats::qnorm(c(0.5, (1 - level) / 2, (1 + level) / 2)) *
            rep(sqrt(object[[est[i]]]$cov[j, j]), 3)
      }
      row.names(out[[i]]) <- object$a_0
      colnames(out[[i]]) <- c("est", "cil", "ciu")
    }
  } else if (is.numeric(contrast)) {
    # check that contrast came in correctly
    if (length(contrast) != length(object$a_0)) {
      stop("length of contrast vector not equal to length of a_0")
    }
    if (!all(contrast %in% c(-1, 1, 0))) {
      stop("contrast should only be -1, 1, or 0")
    }
    for (i in seq_along(est)) {
      out[[i]] <- matrix(NA, nrow = 1, ncol = 3)
      g <- matrix(contrast, nrow = length(contrast))
      p <- matrix(object[[est[i]]]$est, nrow = length(contrast))
      v <- object[[est[i]]]$cov
      thisC <- t(g) %*% p
      thisSe <- sqrt(t(g) %*% v %*% g)
      out[[i]][1, ] <- rep(thisC, 3) +
        stats::qnorm(c(0.5, (1 - level) / 2, (1 + level) / 2)) * rep(thisSe, 3)
      indMinus <- which(contrast == -1)
      indPlus <- which(contrast == 1)
      plusTerms <- minusTerms <- ""
      if (length(indMinus) > 0) {
        minusTerms <- paste0("E[Y(", object$a_0[indMinus], ")]")
      }
      if (length(indPlus) > 0) {
        plusTerms <- paste0("E[Y(", object$a_0[indPlus], ")]")
      }
      thisName <- paste0(
        paste0(plusTerms, collapse = "+"),
        ifelse(length(indMinus) > 0, "-", ""),
        paste0(minusTerms, collapse = "-")
      )

      row.names(out[[i]]) <- thisName
      colnames(out[[i]]) <- c("est", "cil", "ciu")
    }
  } else if (is.list(contrast)) {
    if (!all(c("f", "f_inv", "h", "fh_grad") %in% names(contrast))) {
      stop("some function missing in contrast.")
    }
    for (i in seq_along(est)) {
      out[[i]] <- matrix(NA, nrow = 1, ncol = 3)
      thisC <- do.call(contrast$h, args = list(est = object[[est[i]]]$est))
      f_thisC <- do.call(contrast$f, args = list(eff = thisC))

      grad <- matrix(do.call(
        contrast$fh_grad,
        args = list(est = object[[est[i]]]$est)
      ), nrow = length(object$a_0))
      v <- object[[est[i]]]$cov
      thisSe <- sqrt(t(grad) %*% v %*% grad)
      transformCI <- rep(f_thisC, 3) +
        stats::qnorm(c(0.5, (1 - level) / 2, (1 + level) / 2)) *
          rep(thisSe, 3)
      out[[i]][1, ] <- do.call(contrast$f_inv, args = list(eff = transformCI))
      row.names(out[[i]]) <- c("user contrast")
      colnames(out[[i]]) <- c("est", "cil", "ciu")
    }
  }
  class(out) <- "ci.adaptive_iptw"
  return(out)
}
